r/tests/testthat/test-duckdb.R

```diff
@@ -15,11 +15,25 @@
 # specific language governing permissions and limitations
 # under the License.
 
-skip_if_not_installed("duckdb", minimum_version = "0.3.1")
-skip_if_not_installed("dbplyr")
 skip_if_not_available("dataset")
 skip_on_cran()
 
+# this test needs to be the first one since all other test blocks are skipped
+# if duckdb is not installed
+test_that("meaningful error message when duckdb is not installed", {
+  # skipping if duckdb is installed since we're testing the to_duckdb function's
+  # complaint when a user tries to call it, but duckdb isn't available
+  skip_if(requireNamespace("duckdb", quietly = TRUE))
+  ds <- InMemoryDataset$create(example_data)
+  expect_snapshot(
+    to_duckdb(ds),
+    error = TRUE
+  )
```

Comment:
I'm not opposed to `expect_snapshot()` generally (especially checking the outputs of a bunch of errors in a row, or with particularly complicated messages), but with one single error message like this, it seems like a lot of extra work + redirection for not much benefit. How would you feel about keeping this `expect_error()`?

Reply:
`expect_error()` didn't seem to like the longer error message (it was failing even though the message matched). I could've maybe passed `fixed = TRUE` via the `...` (but I keep forgetting that argument exists).

The `testthat` crew also recommend using `expect_snapshot()` for these sort of use-cases:
> If you are using expect_error() to check that an error message is formatted in such a way that it makes sense to a human, we recommend using [expect_snapshot()](http://127.0.0.1:29237/help/library/testthat/help/expect_snapshot) instead.

I'm not 100% opposed to using `expect_error()`, but I tend to favour snapshots since, at least from my point of view, they're more developer friendly.

Reply:
Having said that, I'd be happy to keep `expect_snapshot()` for the more involved / composed error messages.

Reply:
Done. Switched to `expect_error()`.

Reply:
I read that advice as being if you've got custom formatting (or you're using rlang (i) ... (x) ... format) that `expect_snapshot()` is the way to go (with which I totally agree!). This is really a test that there is an error + it has some of the text we're expecting — there isn't really formatting so to speak so we're not testing that formatting. 

Same with "It can be tricky to test for a combination of different conditions, such as a message followed by an error. [expect_snapshot()](https://testthat.r-lib.org/reference/expect_snapshot.html) is often an easier alternative for these more complex cases." (from https://testthat.r-lib.org/reference/expect_error.html) and since that's not what we're doing here we don't need to have the extra indirection of a separate file, etc.